UI/Pages/Cookies.cshtml.cs

```diff
@@ -7,15 +7,77 @@ namespace UI.Pages
     public class Cookies : PageModel
     {
         private const string ConsentCookieName = ".FindATuitionPartner.Consent";
-        public IActionResult OnGet(bool? consent, string returnUrl)
+        public bool? Consent { get; set; }
+        public bool PreferencesSet { get; set; } = false;
+        public string? returnPath { get; set; }
+
+        public IActionResult OnGet(bool? consent, string returnUrl, bool? viewCookieStatement)
         {
+            returnPath = returnUrl;
+
+            if (Request.Cookies.ContainsKey(ConsentCookieName))
+            {
+                Consent = bool.Parse(Request.Cookies[ConsentCookieName]);
+            }
+
             if (consent.HasValue)
             {
-                var cookieOptions = new CookieOptions { Expires = DateTime.Today.AddMonths(12), Secure = true };
+                PreferencesSet = true;
+                ApplyCookieConsent(consent);
+
+                if (!string.IsNullOrEmpty(returnUrl))
+                {
+                    return Redirect(returnUrl);
+                }
+            }
+            return Page();
+        }
+
+        public IActionResult OnPost(bool? consent, string returnUrl)
+        {
+            returnPath = returnUrl;
+
+            if (Request.Cookies.ContainsKey(ConsentCookieName))
+            {
+                Consent = bool.Parse(Request.Cookies[ConsentCookieName]);
+            }
+
+            if (consent.HasValue)
+            {
+                Consent = consent;
+                PreferencesSet = true;
+
+                var cookieOptions = new CookieOptions { Expires = DateTime.Today.AddMonths(6), Secure = true };
                 Response.Cookies.Append(ConsentCookieName, consent.Value.ToString(), cookieOptions);
-                return new RedirectResult(returnUrl);
+
+                if (!consent.Value)
+                {
+                    ApplyCookieConsent(consent);
+                }
+                return Page();
             }
+
             return Page();
         }
+
+        private void ApplyCookieConsent(bool? consent)
+        {
+            if (consent.HasValue)
+            {
+                var cookieOptions = new CookieOptions { Expires = DateTime.Today.AddMonths(12), Secure = true };
+                Response.Cookies.Append(ConsentCookieName, consent.Value.ToString(), cookieOptions);
+            }
+
+            if (!consent.Value)
```

Comment:
Will this throw an exception if `consent` is null?

Reply:
Null now checked